File: services/audit-service/src/test/java/com/visor/school/audit/model/AuditRecordTest.java
package com.visor.school.audit.model;

import org.junit.jupiter.api.Test;
import java.util.Map;
import java.util.UUID;
import java.util.List;
import java.util.Arrays;

import static org.junit.jupiter.api.Assertions.*;

public class AuditRecordTest {

    @Test
    public void shouldCreateAuditRecordWithRequiredFields() {
        AuditRecord record = new AuditRecord(
            UUID.randomUUID(),
            AuditAction.AUTHENTICATION,
            "User",
            UUID.randomUUID().toString(),
            "192.168.1.1",
            "Mozilla/5.0",
            null,
            true,
            null
        );

        assertNotNull(record.getId());
        assertNotNull(record.getUserId());
        assertEquals(AuditAction.AUTHENTICATION, record.getAction());
        assertEquals("User", record.getResourceType());
        assertNotNull(record.getTimestamp());
        assertEquals("192.168.1.1", record.getIpAddress());
        assertEquals("Mozilla/5.0", record.getUserAgent());
    }

    @Test
    public void shouldCreateAuditRecordWithAllFields() {
        AuditRecord record = new AuditRecord(
            UUID.randomUUID(),
            AuditAction.DATA_MODIFICATION,
            "Student",
            UUID.randomUUID().toString(),
            "192.168.1.1",
            "Mozilla/5.0",
            Map.of("field", "gradeLevel", "oldValue", "9", "newValue", "10"),
            true,
            null
        );

        assertEquals(AuditAction.DATA_MODIFICATION, record.getAction());
        assertEquals("Student", record.getResourceType());
        assertNotNull(record.getDetails());
        assertTrue(record.isSuccess());
        assertNull(record.getErrorMessage());
    }

    @Test
    public void shouldCreateAuditRecordForFailedAction() {
        AuditRecord record = new AuditRecord(
            UUID.randomUUID(),
            AuditAction.ACCESS_ATTEMPT,
            "Student",
            UUID.randomUUID().toString(),
            "192.168.1.1",
            "Mozilla/5.0",
            null,
            false,
            "Access denied: Insufficient permissions"
        );

        assertFalse(record.isSuccess());
        assertEquals("Access denied: Insufficient permissions", record.getErrorMessage());
    }

    @Test
    public void shouldAcceptAllAuditActionTypes() {
        List<AuditAction> actions = Arrays.asList(
            AuditAction.AUTHENTICATION,
            AuditAction.ACCESS_ATTEMPT,
            AuditAction.DATA_MODIFICATION,
            AuditAction.DATA_CREATION,
            AuditAction.DATA_DELETION,
            AuditAction.PASSWORD_RESET,
            AuditAction.EMAIL_VERIFICATION
        );

        actions.forEach(action -> {
            AuditRecord record = new AuditRecord(
                UUID.randomUUID(),
                action,
                "Test",
                UUID.randomUUID().toString(),
                "192.168.1.1",
                "Test",
                null,
                true,
                null
            );

            assertEquals(action, record.getAction());
        });
    }
}

File: services/audit-service/src/test/kotlin/com/visor/school/audit/model/AuditRecordTest.kt
package com.visor.school.audit.model

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import java.util.UUID

class AuditRecordTest {

    @Test
    fun `should create audit record with required fields`() {
        val record = AuditRecord(
            userId = UUID.randomUUID(),
            action = AuditAction.AUTHENTICATION,
            resourceType = "User",
            resourceId = UUID.randomUUID().toString(),
            ipAddress = "192.168.1.1",
            userAgent = "Mozilla/5.0"
        )

        assertNotNull(record.id)
        assertNotNull(record.userId)
        assertEquals(AuditAction.AUTHENTICATION, record.action)
        assertEquals("User", record.resourceType)
        assertNotNull(record.timestamp)
        assertEquals("192.168.1.1", record.ipAddress)
        assertEquals("Mozilla/5.0", record.userAgent)
    }

    @Test
    fun `should create audit record with all fields`() {
        val record = AuditRecord(
            userId = UUID.randomUUID(),
            action = AuditAction.DATA_MODIFICATION,
            resourceType = "Student",
            resourceId = UUID.randomUUID().toString(),
            ipAddress = "192.168.1.1",
            userAgent = "Mozilla/5.0",
            details = mapOf("field" to "gradeLevel", "oldValue" to "9", "newValue" to "10"),
            success = true,
            errorMessage = null
        )

        assertEquals(AuditAction.DATA_MODIFICATION, record.action)
        assertEquals("Student", record.resourceType)
        assertNotNull(record.details)
        assertTrue(record.success)
        assertNull(record.errorMessage)
    }

    @Test
    fun `should create audit record for failed action`() {
        val record = AuditRecord(
            userId = UUID.randomUUID(),
            action = AuditAction.ACCESS_ATTEMPT,
            resourceType = "Student",
            resourceId = UUID.randomUUID().toString(),
            ipAddress = "192.168.1.1",
            userAgent = "Mozilla/5.0",
            success = false,
            errorMessage = "Access denied: Insufficient permissions"
        )

        assertFalse(record.success)
        assertEquals("Access denied: Insufficient permissions", record.errorMessage)
    }

    @Test
    fun `should accept all audit action types`() {
        val actions = listOf(
            AuditAction.AUTHENTICATION,
            AuditAction.ACCESS_ATTEMPT,
            AuditAction.DATA_MODIFICATION,
            AuditAction.DATA_CREATION,
            AuditAction.DATA_DELETION,
            AuditAction.PASSWORD_RESET,
            AuditAction.EMAIL_VERIFICATION
        )

        actions.forEach { action ->
            val record = AuditRecord(
                userId = UUID.randomUUID(),
                action = action,
                resourceType = "Test",
                resourceId = UUID.randomUUID().toString(),
                ipAddress = "192.168.1.1",
                userAgent = "Test"
            )

            assertEquals(action, record.action)
        }
    }
}

